PocketKit/Sources/SharedPocketKit/Migrations/LegacyUserMigration.swift

```diff
@@ -0,0 +1,105 @@
+import Foundation
+
+public enum LegacyUserMigrationError: Error, LocalizedError {
+    case missingKey
+    case missingStore
+    case failedDecryption(Error)
+    case missingData
+    case failedDeserialization(Error)
+
+    public var errorDescription: String? {
+        switch self {
+        case .missingKey: return "Missing decryption key for legacy user migration"
+        case .missingStore: return "Missing file for legacy store"
+        case .failedDecryption(let error): return "Failed to decrypt store: \(error)"
+        case .missingData: return "Data is missing after decrypting store"
+        case .failedDeserialization(let error): return "Failed to deserialize data: \(error)"
+        }
+    }
+}
+
+public class LegacyUserMigration {
+    static let migrationKey = "com.mozilla.pocket.next.migration.legacyUser"
+    static let versionKey = "version"
+    private static let decryptionKey = "kPKTCryptoKey"
+
+    private let userDefaults: UserDefaults
+    private let encryptedStore: LegacyEncryptedStore
+    private let appSession: AppSession
+
+    public var isRequired: Bool {
+        return !previouslyRun && requiredByVersion
+    }
+
+    public init(userDefaults: UserDefaults, encryptedStore: LegacyEncryptedStore, appSession: AppSession) {
+        self.userDefaults = userDefaults
+        self.encryptedStore = encryptedStore
+        self.appSession = appSession
+    }
+
+    public func perform() throws {
+        guard isRequired == true else {
```

Comment:
@dskuza we should put analytics in this code path, that way we can tell when its ok to remove all this.